Framework: Rename `UrlInput` component as `URLInput`

aduth opened the pull request:
This pull request seeks to effect new capitalization guidelines around abbreviations (#7670) toward an API freeze. It deprecates the `wp.editor.UrlInput` component in favor of the corrected `wp.editor.URLInput` naming. 

**Testing instructions:**

Verify that there are no regressions in the use of the URLInput component (e.g. paragraph or button block link input).

Ensure that `wp.editor.UrlInput` works unaffected, but logs a deprecated warning.

aduth committed:
Editor: Deprecate UrlInput as renamed URLInput

aduth committed:
Framework: Remove deprecated UrlInput usage

jorgefilipecosta commented on editor/components/url-input/button.js:
```diff
@@ -76,4 +77,17 @@ class UrlInputButton extends Component {
+		deprecated( 'wp.editor.UrlInputButton', {
+			alternative: 'wp.editor.URLInputButton',
+			plugin: 'Gutenberg',
+			version: 'v3.4',
```
This code will be released on 3.3 should the deprecation version be 3.5?

gziolo replied:
Yes, it should be `3.5`.

jorgefilipecosta reviewed (commented):
I think we may have an error on our deprecation logic. I changed the usage of URLInput on button back to UrlInput for testing purposes and when choosing a link to another post when using UrlInput I see the error:
![image](https://user-images.githubusercontent.com/11271197/42777554-5a3f7582-8932-11e8-887f-7e50e05e313f.png)

gziolo commented on editor/components/url-input/index.js:
```diff
@@ -239,4 +240,17 @@ class UrlInput extends Component {
 }
 
-export default withSpokenMessages( withInstanceId( UrlInput ) );
+export class UrlInput extends URLInput {
```
I think it needs to be exposed also with HOCs applied.

gziolo commented on editor/components/url-input/index.js:
```diff
@@ -239,4 +240,17 @@ class UrlInput extends Component {
+		deprecated( 'wp.editor.UrlInput', {
+			alternative: 'wp.editor.URLInput',
+			plugin: 'Gutenberg',
+			version: 'v3.4',
```
`3.5` again, I'm not quite sure what version should be used in general, but we already have a few entries scheduled for `3.5`: https://github.com/WordPress/gutenberg/blob/master/docs/reference/deprecated.md#350.

We should also add this one to the list.

aduth commented:
- Changed deprecation version to 3.5
- Added deprecation note to `deprecated.md`
- Fixed errors with `UrlInput` shim
   - I abandoned `extends URLInput`, and just made it a pass-through via `render`. I'm still not entirely clear why the previous implementation did not work, though maybe related to @gziolo note on HOC. I don't think we should need them as a pass-through.

jorgefilipecosta reviewed (commented):
I'm seeing an error when adding a link to an existing post in a paragraph without any change: 
![image](https://user-images.githubusercontent.com/11271197/42898868-496cc9a4-8abc-11e8-9671-754ed1172a11.png)

youknowriad commented on editor/components/url-input/index.js:
```diff
@@ -239,4 +240,21 @@ class UrlInput extends Component {
 }
 
-export default withSpokenMessages( withInstanceId( UrlInput ) );
+export class UrlInput extends Component {
```
Should we wrap this one in `withSpokenMessages( withInstanceId(` ?

aduth replied:
>Should we wrap this one in `withSpokenMessages( withInstanceId(` ?

The canonical `URLInput` has these higher-order components still, and this one is just a pass-through that has no functionality of its own, so it shouldn't be necessary.

youknowriad replied:
From my understanding you're not using the `URLInput` that's wrapped with those, unless I'm misreading :)

aduth replied:
Ah, I see now. You're right.

...The dangers of applying higher-order components as part of the `export default` statement 😬

aduth replied:
Should be fixed now. Verified `instanceId` on the passthrough component is assigned into its ID.

youknowriad commented:
>  I have a feeling this is related to the recent api-fetch refactor, changing from a jqXHR type (with abort function) to a Promise (from fetch) which has no cancellation†

Interesting, if it's a common usage to abort apiRequests, we can probably add a custom `abort` function to the returned promise?

aduth commented:
>Interesting, if it's a common usage to abort apiRequests, we can probably add a custom `abort` function to the returned promise?

It shouldn't be a common need. Even here, I don't know that we'd strictly need to cancel the request when the component unmounts, vs. allowing it to continue and just not handling the response.

aduth committed:
Components: Ensure UrlInput passthrough uses enhanced URLInput

youknowriad reviewed (approved):
LGTM 👍
